### PR TITLE
feat(store): implement `ngxsAfterBootstrap` lifecycle hook

### PR DESCRIPTION
**Pull request by arturovt**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

https://github.com/ngxs/store/issues/547
https://github.com/ngxs/store/issues/749

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by splincode** [comment]
Great work!!!

1. Add a test for server side rendering
https://github.com/ngxs/store/blob/master/integration/tests-ssr/todo.mocha.ts

```ts
it('Should be exist "ngxsAfterBootstrap todo"', async () => {
    body = await request('http://localhost:4000/list');
    expect(body.includes('ngxsAfterBootstrap todo')).to.equal(true);
  });
```

2. You must wait for the merge of this PR
https://github.com/ngxs/store/pull/747

Since it is related to the fact that we are doing a module refactor

3. Add unit tests for root and lazy modules

File: packages/store/tests/state.spec.ts [hunk 1]
```diff
@@ -163,4 +170,63 @@ describe('State', () => {
...
+      oldRoots.forEach(oldRoot => adapter.remove(oldRoot));
+
+      adapter.appendChild(document.body, root);
+    }
+
+    @Component({
+      selector: 'app-root',
+      template: ''
+    })
+    class MockComponent {}
```
Comment (splincode) [inline packages/store/tests/state.spec.ts]:
```ts
MockComponent {
  public ngOnInit() {
   hooks.push('ngOnInit');
  }
  
  ..

  public ngAfterViewChecked() {
   hooks.push('ngAfterViewChecked');
  }


}
```

File: packages/store/tests/state.spec.ts [hunk 2]
```diff
@@ -163,4 +170,63 @@ describe('State', () => {
...
+
+      TestBed.configureTestingModule({
+        imports: [createModule(), NgxsModule.forRoot([FooState])]
+      });
+
+      const app: ApplicationRef = TestBed.get(ApplicationRef);
+      createRootNode();
+      app.bootstrap(MockComponent);
+
+      expect(hooks[0]).toBe('ngxsOnInit');
```
Comment (splincode) [inline packages/store/tests/state.spec.ts]:
```ts
expect(hooks).toBe([ 
'ngxsOnInit', 
 ...
 'ngAfterViewChecked',
'ngxsAfterBootstrap'
]);
```

File: packages/store/src/internal/bootstrapper.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { Injectable } from '@angular/core';
+
+import { ReplaySubject, Observable } from 'rxjs';
+
+@Injectable()
+export class Bootstrapper {
+  /**
+   * Use `ReplaySubject`, thus we can get cached value even if the stream is completed
+   */
+  private bootstrap$ = new ReplaySubject<boolean>(1);
```
Comment (splincode) [inline packages/store/src/internal/bootstrapper.ts]: why (1)?

File: packages/store/tests/state.spec.ts [hunk 4]
```diff
@@ -163,4 +170,84 @@ describe('State', () => {
...
+        public ngxsAfterBootstrap(): void {
+          hooks.push(LifecycleHooks.NgxsAfterBootstrap);
+        }
+      }
+
+      TestBed.configureTestingModule({
+        imports: [createModule(), NgxsModule.forRoot([FooState])]
+      });
+
+      const app: ApplicationRef = TestBed.get(ApplicationRef);
```
Comment (splincode) [inline packages/store/tests/state.spec.ts]:
```ts
MockModule implements NgDoBootstrap {
 public static ngDoBootstrap(app: ApplicationRef) {
   createRootNode();
   app.bootstrap(MockComponent);
 }
}
```

```ts
MockModule.ngDoBoostrap(TestBed.get(ApplicationRef))
```

Comment (markwhitfeld) [reply packages/store/tests/state.spec.ts]: Does this lifecyle hook get called for lazy loaded feature modules?

**Comment by markwhitfeld** [comment]
Does this lifecyle hook get called for lazy loaded feature modules?

**Comment by arturovt** [comment]
@markwhitfeld It does, you can see it in the `NgxsFeatureModule`. But application is bootstrapped before any `NgxsFeatureModule` is initialized.

**Comment by splincode** [comment]
@arturovt we must still send the event for those states that have lazily loaded

File: packages/store/src/internal/bootstrapper.ts (hunk 3 above)
Comment (arturovt) [reply packages/store/src/internal/bootstrapper.ts]: @splincode We have to cache 1 value - `true`. Even if this stream is completed - any subscriber will get `true` value if application is bootstrapped. If you don't provide `bufferSize` argument - the default value is `Number.POSITIVE_INFINITY`.

**Comment by arturovt** [comment]
@splincode It is sent :smile:

File: packages/store/src/internal/bootstrapper.ts (hunk 3 above)
Comment (splincode) [reply packages/store/src/internal/bootstrapper.ts]: add comments into code

File: packages/store/src/module.ts [hunk 5]
```diff
@@ -87,11 +92,13 @@ export class NgxsFeatureModule {
       stateOperations.setState({ ...cur, ...results.defaults });
     }
 
-    stateOperations.dispatch(new UpdateState()).subscribe(() => {
-      if (results) {
-        factory.invokeInit(results.states);
-      }
-    });
+    // dispatch the update action and invoke init and bootstrap functions after
+    internalStateOperations.dispatchActionAndInvokeLifecyleHooks(
```
Comment (splincode) [inline packages/store/src/module.ts]: Rename and do as here
https://github.com/ngxs/store/blob/1ccc958b920e13bbd46d6f84976fb56cd00657d0/packages/store/src/module.ts#L40

**Comment by splincode** [comment]
@arturovt add test for init lazy state and invoke ngxsAfterBootstrap

File: packages/store/src/internal/bootstrapper.ts (hunk 3 above)
Comment (arturovt) [reply packages/store/src/internal/bootstrapper.ts]: @splincode where? :wink:

File: packages/store/src/internal/state-factory.ts [hunk 6]
```diff
@@ -155,14 +154,14 @@ export class StateFactory {
    * Invoke the init function on the states.
    */
   invokeInit(stateMetadatas: MappedStore[]) {
```
Comment (splincode) [inline packages/store/src/internal/state-factory.ts]:
```ts
export LifeCycleStateManager {

}
```

0 - rename ngxsBootstrap to invokeNgxsBootstrap
1 - create file: lifecycle-state-manager.ts
2 - move invokeNgxsBootstrap, invokeInit, invokeAfterBootstrap to lifecycle-state-manager.ts

File: packages/store/src/internal/bootstrapper.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    return this.bootstrap$.asObservable();
+  }
+
+  /**
+   * This event will be emitted after attaching `ComponentRef` of the root component
+   * to the tree of views, that's a signal that application has been fully rendered
+   */
+  bootstrap(): void {
+    this.bootstrap$.next(true);
+    this.bootstrap$.complete();
```
Comment (splincode) [inline packages/store/src/internal/bootstrapper.ts]: can I subscribe for use in ngxs hmr plugin?

Comment (arturovt) [reply packages/store/src/internal/bootstrapper.ts]: you can subscribe to `appBootstrapped$` getter, it will emit cached value (if app was bootsrapped)

**Comment by markwhitfeld** [comment]
@splincode Are we going to pull this one in before your other PR for duplicate states?
I'll do my review a little later

**Comment by splincode** [comment]
@markwhitfeld My PR will be implemented later, as there are too many conflicts, I prefer this PR. Because thanks to him we will be able to quickly implement HMR

File: packages/store/src/internal/internals.ts [hunk 8]
```diff
@@ -65,6 +65,13 @@ export interface MappedStore {
   depth: string;
 }
 
+export type StatesAndDefaults =
+  | {
```
Comment (markwhitfeld) [inline packages/store/src/internal/internals.ts]: Is this first `|` supposed to be here? There is nothing before it.

File: packages/store/src/internal/internals.ts [hunk 9]
```diff
@@ -65,6 +65,13 @@ export interface MappedStore {
...
+      defaults: any;
+      states: MappedStore[];
+    }
+  | undefined;
```
Comment (markwhitfeld) [inline packages/store/src/internal/internals.ts]: Why is `undefined` a value in the type, surely specifying the option of undefined at the argument would be a clearer way to express this?
ie. (in lifecycke-state-manager.ts)
`ngxsBootstrap<T>(action: T, results: StatesAndDefaults | undefined): void {`

File: packages/store/src/internal/state-factory.ts [hunk 10]
```diff
@@ -214,7 +199,7 @@ export class StateFactory {
   /**
    * Create the state context
    */
-  private createStateContext(metadata: MappedStore) {
+  createStateContext(metadata: MappedStore) {
```
Comment (markwhitfeld) [inline packages/store/src/internal/state-factory.ts]: This should be kept private and `StateContextFactory` should be injected and used where this was being used.
( added comment to usage in lifecycle-state-manager.ts )

File: packages/store/src/internal/lifecycle-state-manager.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import { Bootstrapper } from './bootstrapper';
+import { InternalStateOperations } from './state-operations';
+import { MappedStore, StatesAndDefaults } from './internals';
+import { LifecycleHooks, NgxsLifeCycle } from '../symbols';
+
+@Injectable()
+export class LifecycleStateManager {
+  constructor(
+    private internalStateOperations: InternalStateOperations,
+    private factory: StateFactory,
```
Comment (markwhitfeld) [inline packages/store/src/internal/lifecycle-state-manager.ts]:
```suggestion
    private stateContextFactory: StateContextFactory,
```
You should rather inject and use `StateContextFactory` here. This is the correct factory to use to create a StateContext.

File: packages/store/src/internal/state-factory.ts (hunk 10 above)
Comment (markwhitfeld) [reply packages/store/src/internal/state-factory.ts]: Also, after moving the `invokeInit` method out of this class the only usage of this method is in the `invokeActions` method. We could actually remove this private method here and call `this._stateContextFactory.createStateContext` directly in the `invokeActions` method.

File: packages/store/src/internal/lifecycle-state-manager.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  invokeBootstrap(stateMetadatas: MappedStore[]) {
+    this.invokeLifecycleHooks(stateMetadatas, LifecycleHooks.NgxsAfterBootstrap);
+  }
+
+  private invokeLifecycleHooks(stateMetadatas: MappedStore[], hook: LifecycleHooks): void {
+    for (const metadata of stateMetadatas) {
+      const instance: NgxsLifeCycle = metadata.instance;
+
+      if (instance[hook]) {
+        const stateContext = this.factory.createStateContext(metadata);
```
Comment (markwhitfeld) [inline packages/store/src/internal/lifecycle-state-manager.ts]:
```suggestion
        const stateContext = this.stateContextFactory.createStateContext(metadata);
```

File: packages/store/tests/state.spec.ts [hunk 13]
```diff
@@ -163,4 +170,109 @@ describe('State', () => {
...
+
+      expect(hooks).toEqual([
+        LifecycleHooks.NgxsOnInit,
+        LifecycleHooks.NgOnInit,
+        LifecycleHooks.NgAfterViewInit,
+        LifecycleHooks.NgxsAfterBootstrap
+      ]);
+    });
+
+    it('should invoke "ngxsAfterBootstrap" for lazy states', () => {
```
Comment (markwhitfeld) [inline packages/store/tests/state.spec.ts]:
```suggestion
    it('should invoke "ngxsAfterBootstrap" for feature states', () => {
```
This is more a test for feature states. Testing lazy states would require a whole dance with the router.

Comment (markwhitfeld) [reply packages/store/tests/state.spec.ts]: Ocurrences of lazy in the test would have to change too ;-)

File: packages/store/src/internal/lifecycle-state-manager.ts (hunk 11 above)
Comment (arturovt) [reply packages/store/src/internal/lifecycle-state-manager.ts]: @markwhitfeld Oh, nice catch! :wink:

File: packages/store/src/internal/internals.ts (hunk 8 above)
Comment (arturovt) [reply packages/store/src/internal/internals.ts]: @markwhitfeld This is prettier :(

**Comment by arturovt** [comment]
@markwhitfeld ping

**Comment by eranshmil** [comment]
Great job, @arturovt! 👍

**Pull request merged**